### PR TITLE
Implement `getCanonicalLocales` in non apple/android environments

### PR DESCRIPTION
**Pull request by manuelpuyol**
Related to https://github.com/facebook/hermes/issues/1350 and https://github.com/facebook/hermes/discussions/1211

## Summary

As an initial step for `Intl` backed by ICU, I'm implementing the `getCanonicalLocales` function using BCP47Parser. The implementation is exactly the same as Apple's implementation

## Test Plan

This change should be covered by the `get-canonical-locales.js` test.

### CODE REVIEW

**Comment by neildhar** [comment]
Hey @manuelpuyol, thanks for putting this together! The main high level suggestion I have here is to use the existing unicode tag parsing and canonicalisation functionality we use for the Apple Intl implementation. In particular, take a look at `BCP47Parser.cpp`, and how it is used in our existing Apple implementation in `PlatformIntlApple.mm`. You may be able to just copy the implementation over from the Apple platform. (we're planning to unify these in the longer term)

That should also solve the ICU issue you shared, since all the tag parsing and canonicalisation will be done by us.

**Comment by manuelpuyol** [comment]
hey @neildhar! I used Apple's implementation and reverted the ICU change.
I'm not sure what is going on with  `test-e2e` though, `compileReleaseKotlin` is failing but doesn't look related to my changes and I see it failing on other PRs

File: lib/Platform/Intl/CMakeLists.txt [hunk 1]
```diff
@@ -22,6 +22,10 @@ if(HERMES_ENABLE_INTL)
     set_target_properties(hermesPlatformIntl PROPERTIES UNITY_BUILD false)
     target_compile_options(hermesPlatformIntl PRIVATE -fobjc-arc)
   else()
-    add_hermes_library(hermesPlatformIntl STATIC PlatformIntlDummy.cpp LINK_LIBS hermesPublic)
+    add_hermes_library(hermesPlatformIntl STATIC PlatformIntlICU.cpp
+        LINK_LIBS
+        hermesBCP47Parser
+        hermesPublic
+    )
```
Comment (neildhar) [inline lib/Platform/Intl/CMakeLists.txt:29]: We need to make sure we link against ICU

```suggestion
    add_hermes_library(hermesPlatformIntl STATIC PlatformIntlICU.cpp
        LINK_LIBS
        hermesBCP47Parser
        hermesPublic
    )
    hermes_link_icu(hermesPlatformIntl)
```

**Review by neildhar — COMMENTED** [review]
LGTM, looking at your other diff, it looks like `getCanonicalLocales` might belong in `PlatformIntlShared.cpp`, but I'll leave it up to you.

We should also tweak the logic here to be more robust:
```
```

File: lib/Platform/Intl/CMakeLists.txt (hunk 1 above)
Comment (manuelpuyol) [reply lib/Platform/Intl/CMakeLists.txt:29]: we aren't using anything from ICU in this implementation, should we still link it?

Comment (neildhar) [reply lib/Platform/Intl/CMakeLists.txt:29]: I suppose not in this particular PR. If you'd prefer to do it in the next PR once ICU is being used, we can just do that.

Comment (manuelpuyol) [reply lib/Platform/Intl/CMakeLists.txt:29]: cool, I'm linking it in the DateTimeFormat PR so we should be good :)

File: lib/Platform/Intl/PlatformIntlICU.cpp [hunk 2]
```diff
@@ -16,19 +17,63 @@ using namespace ::hermes;
...
+namespace {
 
+// https://tc39.es/ecma402/#sec-canonicalizelocalelist
+vm::CallResult<std::vector<std::u16string>> canonicalizeLocaleList(
+    vm::Runtime &runtime,
+    const std::vector<std::u16string> &locales) {
+  // 1. If locales is undefined, then a. Return a new empty list
+  if (locales.empty()) {
+    return std::vector<std::u16string>{};
+  }
```
Comment (neildhar) [inline lib/Platform/Intl/PlatformIntlICU.cpp:29]: I'm about to merge this and just wanted to note two minor things to fix up. 

1. We use `///` for doc-comments, and it is generally preferable to use links to a specific version of the spec (as is done in the Apple version).
2. It looks like there's an extra line here that isn't in the Apple version. Note that given how the function is written, this check is not needed.

```suggestion
/// https://402.ecma-international.org/8.0/#sec-canonicalizelocalelist
vm::CallResult<std::vector<std::u16string>> canonicalizeLocaleList(
    vm::Runtime &runtime,
    const std::vector<std::u16string> &locales) {
  // 1. If locales is undefined, then
  //   a. Return a new empty List
  // Not needed, this validation occurs closer to VM in 'normalizeLocales'.
  // 2. Let seen be a new empty List.
  std::vector<std::u16string> seen;
```

**Pull request closed without merging**